### PR TITLE
Future dates cause errors in filterable lists

### PR DESCRIPTION
**Pull request by cwdavies**
This change should fix an internal server error seen recently in issue [5365](https://github.local/CFGOV/cfgov-alerts/issues/5365), that has something to do with how filterable lists handle future dates.
Putting future dates in the `from_date` and `to_date` filterable list fields should not cause an internal server error.

## Additions

- Add default_input_formats to `FilterableDateField` class in `cfgov/v1/forms.py`
- Add [freezegun](https://pypi.org/project/freezegun/) and test for future dates

## Removals

- `cfgov/v1/util/date_filter.py` module

## Changes

- `cfgov/v1/forms.py`
- `cfgov/v1/tests/test_forms.py`
- `cfgov/v1/tests/util/test_datetimes.py`
- `cfgov/v1/util/datetimes.py`

## How to test this PR

Attempt to add future date into the FilterableList form using a Safari or IE11 browser
e.g. from_date and to_date: 05/04/2021 and 11/04/2021
Try flipping the month and day, European style, which looks like a time range in the future
Filter a filterable list on a date range entirely in the future

1. Visit https://www.consumerfinance.gov/about-us/newsroom/ in Safari or IE11 in Sauce Labs, or any other browser that doesn't support the native date input control
2. Enter some date filter and click Apply Filters
3. Verify there is no internal server error

## Notes and todos

- This must be testing in IE11, Safari, and Opera Mini browsers

## Checklist

- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

#### Browser testing

Visually tested in the following supported browsers:
- [ ] Firefox
- [x] Chrome
- [x] Safari
- [ ] Edge 18 (the last Edge prior to it switching to Chromium)
- [ ] Internet Explorer 11 and 8 (via emulation in 11's dev tools)
- [ ] Safari on iOS
- [ ] Chrome on Android

#### Accessibility

- [ ] Keyboard friendly (navigable with tab, space, enter, arrow keys, etc.)
- [ ] Screen reader friendly
- [ ] Does not introduce new errors or warnings in [WAVE](https://wave.webaim.org/extension/)

#### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Does not introduce new lint warnings
- [ ] Flexible from small to large screens

### CODE REVIEW

File: cfgov/v1/forms.py [hunk 1]
```diff
@@ -241,11 +244,8 @@ def set_interpreted_date_values(self, cleaned_data):
                     earliest_results, self.preferred_datetime_format)
 
             if to_date:
-                transformed_to_date = end_of_time_period(
-                    self.data['to_date'], cleaned_data['to_date'])
-                cleaned_data['to_date'] = transformed_to_date
                 self.data['to_date'] = date.strftime(
-                    transformed_to_date, self.preferred_datetime_format)
+                    cleaned_data['to_date'], self.preferred_datetime_format)
```
Comment (willbarton) [inline cfgov/v1/forms.py:248]: It looks like [`end_of_time_period` was added to handle ](https://github.com/cfpb/consumerfinance.gov/commit/624c5ae4ad2dddf0b30bb671a3467cc08d017181#diff-40757824786a8c38156d12b1db8227629193ded9e4279342230a81d3e4844250) ambiguous dates. Are those still handled?

Comment (cwdavies) [reply cfgov/v1/forms.py:248]: @willbarton I can add more tests to check for ambiguous dates to verify that they are handled.
Do you have any specific examples in mind?

Comment (cwdavies) [reply cfgov/v1/forms.py]: Good catch. I added `end_of_time_period` to handle ambiguous dates.

**Review by willbarton — CHANGES_REQUESTED** [review]
I can't speak for anyone else, but I often use tests to understand intention behind code. Could you add some tests that specifically test `date_from_pattern` and `end_of_time_period` directly? I suspect re-using the tests for the old functions would be a good way to ensure they continue to work as intended.

**Comment by Scotchester** [comment]
> I can't speak for anyone else, but I often use tests to understand intention behind code. Could you add some tests that specifically test `date_from_pattern` and `end_of_time_period` directly? I suspect re-using the tests for the old functions would be a good way to ensure they continue to work as intended.

+1 to that, and also Craig, would you mind un-renaming the module so that we can clearly see in the diff what has changed from before?

**Comment by cwdavies** [comment]
The GHE issue 4035 stated:
```
From that investigation, we thought the best way to remedy this is to replace the whole `date_filter.py` module
```
That is the reason why I removed both date_filter.py and test_date_filter.py.

I put both `date_from_pattern` and `end_of_time_period` into the `datetimes.py` module.
The above two are still needed to handle ambiguous dates.
My understanding of ambiguous dates are those that contain just a year or just a month and a year.

The original behavior is left intact to use the end of month if only a month and year are entered.
e.g. if a user entered 3/2021 in a filter form, the date used will be March 31st 2021

Also the original behavior is left intact to use the end of year if only a year is entered.
e.g. if a user entered 2020 in a filter form, the date used will be December 31st 2020

All date and time utilities used are now in a single module.

**Review by willbarton — DISMISSED** [review]

**Comment by willbarton** [comment]
@cwdavies The issue was specifically referencing seeing if we could replace it with the use of [`dateutil.parser`](https://dateutil.readthedocs.io/en/stable/parser.html). This was discussed when triaging the issue, but I think in retrospect is probably not possible. Either way, I think this works — the core issue was the returning of a `None` from those functions that we didn't handle in the form itself, and that is definitely fixed.

**Review by willbarton — APPROVED** [review]

**Pull request merged**